### PR TITLE
(proposal) JAMES-2040 MailImpl should not allow duplicated mailAddresses

### PR DESCRIPTION
**Pull request by chibenwa**
I think it will not be enough to solve the issue, but it can be a cool first step: cf integration test...

```
                 2
POSTFIX ------> ML server
           \ \  <-------
             \ \     3 
            1  \ \  4
                 \-\-- > James
```

Why it will not be enough?

 0. I send a mail to op-james@lng.com and to myself.
 1. Postfix server sends email to James
 2. As a ML is also present we need to expend it. It's thus sent to ML server...
 3. Which resent it to postfix, mentionning my name...
 4. s it's traited as a new mail, POSTFIX resend it to James, in a separated SMTP set of commands....

Why it's usefull?
 - We proove that we handle some case that was not managed before cf integration tests.

How to solve this fully?
 - Store SHA of already received emails, per user? Thus we avoid multiple receptions... (Data-api project, and we can provide jpa, memory and Cassandra implems)

This PR also brings:
 - Better testing for MailImpl
 - Solved NPE for getMessageSize and recipients

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[Building Java 8 aj-169-51ff6909-9e0e-458e-a0af-2d992d158693] Failed tests: 
[Building Java 8 aj-169-51ff6909-9e0e-458e-a0af-2d992d158693]   CassandraJmapAuthenticationTest>JMAPAuthenticationTest.deleteMustInvalidAuthorizationOnCorrectAuthorization:448 1 expectation failed.
[Building Java 8 aj-169-51ff6909-9e0e-458e-a0af-2d992d158693] Expected status code <401> doesn't match actual status code <200>.
```

test this please

File: server/container/core/src/main/java/org/apache/james/core/MailImpl.java [hunk 1]
```diff
@@ -411,7 +413,7 @@ public void setMessage(MimeMessage message) {
      */
     @Override
     public void setRecipients(Collection<MailAddress> recipients) {
-        this.recipients = new ArrayList<MailAddress>(recipients);
+        this.recipients = new HashSet<MailAddress>(recipients);
```
Comment (rouazana) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: Question: do you think keeping the order should be of any interest?

Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: No idea... 

Maybe a TreeSet should be more appropriate then, let's be conservative about the order...

Comment (rouazana) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: You mean a LinkedHastSet to have the same order than the add order?

Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: Yes, sorry...

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 2]
```diff
@@ -18,37 +18,34 @@
...
-import java.io.ByteArrayInputStream;
-import java.util.ArrayList;
-import static org.junit.Assert.*;
+import org.apache.mailet.Mail;
+import org.apache.mailet.MailAddress;
+import org.apache.mailet.base.test.MailUtil;
 import org.junit.Test;
 
-public class MailImplTest extends MailTestAllImplementations {
+public class MailImplTest {
```
Comment (strokyl) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: If this is not a implementation of any abstract class/interface why the "Impl" in "MailImplTest" I would rename it to "MailTest"

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 3]
```diff
@@ -18,37 +18,34 @@
...
 
     private static final Session NO_SESSION = null;
-    
-    @Override
-    protected Mail createMailImplementation() {
-        return new MailImpl();
-    }
 
     @Test
     public void testConstr1() throws MessagingException {
```
Comment (strokyl) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: testConstr1 ?

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 4]
```diff
@@ -18,37 +18,34 @@
...
         MailImpl mail = new MailImpl();
 
         helperTestInitialState(mail);
         helperTestMessageSize(mail, 0); // MimeMessageWrapper default is 0
-        assertNull("no initial message", mail.getMessage());
-        assertNull("no initial sender", mail.getSender());
-        assertNull("no initial name", mail.getName());
+
+        assertThat(mail.getMessage()).as("Checking there is an initial message").isNull();
```
Comment (strokyl) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: Don't you mean the contrary "Checking there is No initial message".
Same for the others lines?

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 5]
```diff
@@ -61,12 +58,22 @@ public void testConstr2() throws MessagingException {
...
-        assertNull("no initial message", mail.getMessage());
-        assertEquals("sender", sender, mail.getSender().toString());
-        assertEquals("name", name, mail.getName());
+        assertThat(mail.getMessage()).as("No initial message when not specified").isNull();
+        assertThat(mail.getSender().asString()).as("Sender should match original value").isEqualTo(sender);
+        assertThat(mail.getName()).as("Name should be the specified one").isEqualTo(name);
+    }
+
+    @Test
+    public void setMessageShouldaddAMessage() throws MessagingException {
```
Comment (strokyl) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: `setMessageShouldaddAMessage` -> `setMessageShouldAddAMessage`

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 6]
```diff
@@ -80,31 +87,173 @@ public void testConstr3() throws MessagingException {
...
     }
 
     @Test
-    public void testDuplicateNewName() throws MessagingException {
+    public void duplicateShouldNotChangeOriginalMailName() throws MessagingException {
         String newName = "aNewName";
-
         MailImpl mail = new MailImpl();
-        assertFalse("before + after names differ", newName.equals(mail.getName()));
+        assertThat(mail.getName()).as("before + after names should differ").isNotEqualTo(newName);
```
Comment (strokyl) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: Sorry but I reallly do not understand this test, why would it fail ?

File: server/container/core/src/main/java/org/apache/james/core/MailImpl.java [hunk 7]
```diff
@@ -139,11 +140,9 @@ public MailImpl(String name, MailAddress sender, Collection<MailAddress> recipie
         this();
         this.name = name;
         this.sender = sender;
-        this.recipients = null;
 
         // Copy the recipient list
         if (recipients != null) {
```
Comment (strokyl) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: I think testing for != null is now useless

File: server/container/core/src/main/java/org/apache/james/core/MailImpl.java [hunk 8]
```diff
@@ -125,6 +125,7 @@ public MailImpl() {
         setState(Mail.DEFAULT);
         attributes = new HashMap<String, Object>();
         perRecipientSpecificHeaders = new PerRecipientHeaders();
+        recipients = new ArrayList<MailAddress>();
```
Comment (strokyl) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: I think recipients should be final now, no ?

File: server/container/core/src/main/java/org/apache/james/core/MailImpl.java [hunk 9]
```diff
@@ -541,6 +542,7 @@ private void writeObject(java.io.ObjectOutputStream out) throws IOException {
         out.writeObject(remoteAddr);
         out.writeObject(lastUpdated);
         out.writeObject(attributes);
+        out.writeObject(perRecipientSpecificHeaders);
```
Comment (strokyl) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: Stupid question why here we don't just do `out.writeObject(this)` and annotate with `@transient` all the field we do not want ?

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 10]
```diff
@@ -243,4 +250,22 @@ protected void helperTestInitialState(Mail mail) {
         assertThat(mail.getState()).as("Expecting default initial state").isEqualTo(Mail.DEFAULT);
     }
 
+    @Test
+    public void mailImplShouldBeSerializable() throws Exception {
+        MailImpl initialMail = new MailImpl();
+        initialMail.setRecipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES));
+        initialMail.addSpecificHeaderForRecipient(
+            PerRecipientHeaders.Header.builder()
```
Comment (strokyl) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: It look like you don't check for a lot of property of mail that should be serialized.

File: server/container/core/src/main/java/org/apache/james/core/MailImpl.java [hunk 11]
```diff
@@ -125,7 +127,7 @@ public MailImpl() {
         setState(Mail.DEFAULT);
         attributes = new HashMap<String, Object>();
         perRecipientSpecificHeaders = new PerRecipientHeaders();
-        recipients = new ArrayList<MailAddress>();
+        recipients = new HashSet<MailAddress>();
```
Comment (strokyl) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: Are you sure order does not matter ? Before we did keep order of insertion.

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DuplicatedRecipients.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            .addProcessor(CommonProcessors.spam())
+            .addProcessor(CommonProcessors.localAddressError())
+            .addProcessor(CommonProcessors.relayDenied())
+            .addProcessor(CommonProcessors.bounces())
+            .addProcessor(CommonProcessors.sieveManagerCheck())
+            .build();
+
+        jamesServer = new TemporaryJamesServer(temporaryFolder, mailetContainer);
+        Duration slowPacedPollInterval = Duration.FIVE_HUNDRED_MILLISECONDS;
+        calmlyAwait = Awaitility.with().pollInterval(slowPacedPollInterval).and().with().pollDelay(slowPacedPollInterval).await();
```
Comment (strokyl) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DuplicatedRecipients.java]: Break line please

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java (hunk 6 above)
Comment (chibenwa) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: Don't know.

However, it was here before. There might be a reason. Even if it looks *obvious*. Thus I should not delete here in this commit (simply about test refactoring) but in another commit (like the NullPointerException)

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 13]
```diff
@@ -80,31 +87,173 @@ public void testConstr3() throws MessagingException {
...
+    }
+
+    protected void helperTestInitialState(Mail mail) {
+        assertThat(mail.hasAttributes()).as("Expecting no initial attributes").isFalse();
+        assertThat(mail.getErrorMessage()).as("Expecting no initial error").isNull();
+        assertThat(mail.getLastUpdated()).as("Expecting initial last update set").isNotNull();
+        try {
+            mail.getRecipients().isEmpty();
+        } catch (NullPointerException e) {
+            // current behavior. *BUT*, shouldn't this method better return with
```
Comment (chibenwa) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: Agree. See next commits. 

Upgrade code quality first.

Fix and change behaviour later.

File: server/container/core/src/main/java/org/apache/james/core/MailImpl.java (hunk 9 above)
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/core/MailImpl.java]: My understanding is the previous writer wants this serialization to be as stable as possible.

They gained more stability (across class versions) by specifying manually the fields to write.

Anyway, I also checked, and this serialisation code is not used anywhere...

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java [hunk 14]
```diff
@@ -80,31 +87,173 @@ public void testConstr3() throws MessagingException {
...
-
         MailImpl mail = new MailImpl();
-        assertFalse("before + after names differ", newName.equals(mail.getName()));
+        assertThat(mail.getName()).as("before + after names should differ").isNotEqualTo(newName);
 
-        MailImpl duplicate = (MailImpl) mail.duplicate(newName);
-        assertEquals("new name set", newName, duplicate.getName());
-        helperTestInitialState(duplicate);
-        helperTestMessageSize(duplicate, 0);
+        mail.duplicate(newName);
```
Comment (quynhn) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: setting in the middle of assert part?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DuplicatedRecipients.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @After
+    public void tearDown() {
+        jamesServer.shutdown();
+    }
+
+    @Test
+    public void simpleMailShouldBeSent() throws Exception {
```
Comment (quynhn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DuplicatedRecipients.java]: mailShouldNotDuplicatedWhenDuplicatedRecipient?

File: server/container/core/src/test/java/org/apache/james/core/MailImplTest.java (hunk 14 above)
Comment (chibenwa) [inline server/container/core/src/test/java/org/apache/james/core/MailImplTest.java]: That's the way this test was initially written. 

As noted by Luc it do not test much things. The assertion seems useless, so I removed it.

**Comment by chibenwa** [comment]
test this please

**Pull request closed without merging**